Common/Final.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Common
+{
+	/// <summary>
+	/// This class enforces one-time only initialization.
+	/// Normally you want to use `readonly` for that, but sometimes you can't
+	/// since you're not setting the value from inside a constructor.
+	/// </summary>
+	/// <typeparam name="TItem">The type of the stored value</typeparam>
+	public class Final<TItem>
+	{
+		private bool isSet = false;
+		private TItem value;
+		public TItem Value
+		{
+			get
+			{
+				if(!isSet)
+					throw new ObjectNotSet();
+				return value;
+			}
+			set
+			{
+				if(isSet)
+					throw new ObjectAlreadySet();
```

Comment:
Nothing wrong with it but custom exceptions are rarely needed ime.
I think `InvalidOperationException` with a descriptive message is more useful. If you need to switch and filter a custom exception probably makes sense but this does not look like an exception that should ever be handled.